### ISSUE TITLE
[BUG] Kotlin-Spring Type mismatch for post -> required)true

### ISSUE THREAD

**Issue by alexbulyha** [OpenAPITools/openapi-generator#10832]
#### Bug Report Checklist

- [x] Have you provided a full/minimal spec to reproduce the issue?
- [x] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [ ] Have you [tested with the latest master](https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-to-test-with-the-latest-master-of-openapi-generator) to confirm the issue still exists?
- [x] Have you searched for related issues/PRs?
- [x] What's the actual output vs expected output?
- [ ] [Optional] Sponsorship to speed up the bug fix or feature request ([example](https://github.com/OpenAPITools/openapi-generator/issues/6178))

##### Description

(The generation works perfectly fine for typescript-angular)

The generated post request expects a "SearchRequest" object, the generated test provides a "SearchRequest?", thus failing

SearchApiTest.kt:[24,75] Type mismatch: inferred type is SearchRequest? but SearchRequest was expected
```
@Test
    fun searchPostTest() {
        val searchRequest:SearchRequest? = null
        val response: ResponseEntity<List<SearchResult>> = api.searchPost(searchRequest)

        // TODO: test validations
    }



    fun searchPost( @Valid @RequestBody searchRequest: SearchRequest
): ResponseEntity<List<SearchResult>> {
        return ResponseEntity(HttpStatus.NOT_IMPLEMENTED)
    }
```

Since we use the generated code as a library to be imported in other projects, it's an inconvenience

##### openapi-generator version

openapi: 3.0.1
penapitools/openapi-generator-cli:v5.3.0

##### OpenAPI declaration file content or url

```yaml
openapi: 3.0.1
info:
  title: Search
  version: v1
servers:
  - url: /api/v1
paths:
  /search:
    post:
      requestBody:
        required: true
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/SearchRequest'
      responses:
        "200":
          description: Ad Spaces found
          content:
            application/json:
              schema:
                items:
                  $ref: '#/components/schemas/SearchResult'
                type: array
components:
  schemas:
    SearchRequest:
      type: object
      properties:
        types:
          type: string
    SearchResult:
      type: object
      properties:
        name:
          type: string
```

##### Generation Details
image: openapitools/openapi-generator-cli:v5.3.0
docker-entrypoint.sh generate -i ./openapi/openapi.yaml -g kotlin-spring -o ./build --additional-properties=artifactVersion=${VERSION} --additional-properties=artifactId=api-server

##### Suggest a fix

In case of required, don't use "null" but instead create an empty object? 

I hope there is something i'm missing or a quickfix besides skipping tests!

Thanks and best regards

**Comment by agilob** [comment]
>In case of required, don't use "null" but instead create an empty object?

No, empty object cannot be used, you are confusing `required` and `nullable`. required means that the field **has to be specified** even it is null. If you want object that would default to something - like empty object - you need to make the object not null. Experiment with not null on field, or required *and* not null. The generated output is correct and compatible with other languages what support null safe objects.

**Comment by alexbulyha** [comment]
Hi agilob,

thank you for the explanation!

However, I wouldn't call the generated output "correct" if it throws (unmodified) a Type Mismatch exception.

Maybe I should try to rephrase the issue:

The generated function for the API is correct:
```fun searchPost( @Valid @RequestBody searchRequest: SearchRequest)```


The generated function for the API Test is not:
```
val searchRequest:SearchRequest? = null
val response: ResponseEntity<List<SearchResult>> = api.searchPost(searchRequest)
```

The issue here is, the API expects an Instance of "SearchRequest", while the test provides an Instance of "SearchRequest?" -> Even though it looks very similar, it doesnt work.


### Explanation of my proposed solution:

Why did I mention the "null" solution?

The generated API function expect a "SearchRequest"
```fun searchPost( @Valid @RequestBody searchRequest: SearchRequest)```

In Kotlin,
"SearchRequest" and "SearchRequest?" are *not* the same. The left one can NEVER be null, the right one can.

This means:
``` val searchRequest:SearchRequest? = null ``` is valid Kotlin code
``` val searchRequest:SearchRequest = null ``` is NOT valid Kotlin code


So if we try to just set the correct type (remove "?"):
```
 fun searchPostTest() {
        val searchRequest:SearchRequest = null
        val response: ResponseEntity<List<SearchResult>> = api.searchPost(searchRequest)

        // TODO: test validations
    }
```
It will throw an error:
Null can not be a value of a non-null type SearchRequest


The only way to satisfy the correct Type, "SearchRequest" is by 
1. Changing the declaration to "SearchRequest" (removing ?)
2. Providing a valid non-null Object

Unfortunately Kotlins method to remove Javas NullPointer Exceptions causes this other issues




### How does this relate to "required"?

If you remove the required attribute, the generated function for the API looks like this:
```
    fun searchPost( @Valid @RequestBody(required = false) searchRequest: SearchRequest?
): ResponseEntity<List<SearchResult>> {
        return ResponseEntity(HttpStatus.NOT_IMPLEMENTED)
    }
```

Here you can see, it accepts "SearchRequest?"

So the issue is caused be the required function.


### TL;DR

The API is generated correctly.the TEST is not.
The TEST is generated exactly the same no matter the value of the required flag, which leads to issues.
The TEST should be adjusted

**Comment by alexbulyha** [comment]
The same applies to required path parameters

```yaml

    parameters:
      - name: id
        in: path
        required: true
        description: the Entity identifier
        schema:
          type: integer
          format: int64
```

In this case, the generated Api Function expects an id type Long,
meanwhile the test provides a "Long?"

```
    fun entityIdSubEntitiesGet( @PathVariable("id") id: kotlin.Long): ResponseEntity<List<SubEntity>> {
        return ResponseEntity(HttpStatus.NOT_IMPLEMENTED)
    }


@Test
    fun entityIdSubEntitiesGetTest() {
        val id:kotlin.Long? = null
        val response: ResponseEntity<List<SubEntity>> = api.entityIdSubEntitiesGet(id)

        // TODO: test validations
    }

```
Error:

Type mismatch: inferred type is Long? but Long was expected

**Comment by siavashingka** [comment]
I have the same problem, I want to have a `List<Object>?` and I specified that the request parameter is nullable and not required but in the generated code I get `List<Object>`
